view.go

```diff
@@ -179,6 +220,137 @@ func (v *View) Read(ctx context.Context, key string, each func(entry *ViewEntry)
 	return nil
 }
 
+func (v *View) queryMap(ctx context.Context, p ViewQueryParam) (ViewResult, error) {
+	var orderBy string
+	if p.Descending {
+		s := p.StartKey
+		p.StartKey = p.EndKey
+		p.EndKey = s
+		orderBy = "key DESC, doc_id DESC"
+	} else {
+		orderBy = "key ASC, doc_id ASC"
+	}
+
+	ret := ViewResult{}
+	if p.Key != "" && len(p.Keys) > 0 {
+		return ret, fmt.Errorf("cannot supply both key and keys parameter key: %s keys: %v", p.Key, p.Keys)
+	}
+
+	qb := newQueryBuilder()
+	qb.Add(`SELECT key, doc_id, value FROM ` + v.mapperTable + ` WHERE deleted != TRUE`)
+	qb.AddIfNotZero(` AND key = $1`, p.Key)
+	if len(p.Keys) > 0 {
+		if p.Keys[0] == "" {
+			return ret, fmt.Errorf("parameter keys[0] cannot be empty string")
+		}
+		qb.Add(` AND (key = $1`, p.Keys[0])
+		for _, k := range p.Keys[1:] {
+			qb.AddIfNotZero(` OR key = $1`, k)
+		}
+		qb.Add(`)`)
+	}
+	qb.AddIfNotZero(` AND key >= $1`, p.StartKey)
+	eqSign := "="
+	if p.ExclusiveEnd {
+		eqSign = ""
+	}
+	qb.AddIfNotZero(` AND key <`+eqSign+` $1`, p.EndKey)
+	qb.AddIfNotZero(` AND doc_id >= $1`, p.StartKeyDocID)
+	qb.AddIfNotZero(` AND doc_id <= $1`, p.EndKeyDocID)
+
+	qb.Add(" ORDER BY " + orderBy)
+	qb.AddIfNotZero(" LIMIT $1", p.Limit)
+	qb.AddIfNotZero(" OFFSET $1", p.Skip)
+
+	rows, err := qb.Query(ctx, v.sqldb)
+	if err != nil {
+		return ret, err
+	}
+	for rows.Next() {
+		r := ViewResultRow{}
+		if err := rows.Scan(&r.Key, &r.ID, &r.Value); err != nil {
+			return ret, err
+		}
+		ret.Rows = append(ret.Rows, r)
+	}
+
+	r := v.sqldb.QueryRowContext(ctx, `SELECT count(*) FROM `+v.mapperTable+` WHERE deleted != TRUE`)
+	if err := r.Scan(&ret.TotalRows); err != nil {
+		return ret, err
+	}
+	ret.Offset = p.Skip
+
+	return ret, nil
+}
+
+// Query queries the view.
+func (v *View) Query(ctx context.Context, p ViewQueryParam) (ViewResult, error) {
+	r := ViewResult{}
+	var err error
+
+	if p.Stale != "ok" && p.Stale != "update_after" {
+		err = v.Refresh(ctx)
+		if err != nil {
+			return r, err
+		}
+	}
+
+	r, err = v.queryMap(ctx, p)
+	if err != nil {
+		return r, err
+	}
+
+	if p.UpdateSeq {
+		seqStore := v.metaStore.At(v.config.Input.Table.name).At("last_seq")
+		var lastSeq int64
+		err := seqStore.Get(ctx, &lastSeq)
+		if err != nil && !IsError(err, ErrNotFound) {
+			return r, err
+		}
+		r.UpdateSeq = lastSeq
+	}
+
+	if p.NoReduce || v.config.Reducer == nil {
+		for i := range r.Rows {
+			if p.IncludeDocs {
+				r.Rows[i].Doc = &Document{}
+				err = v.config.Input.Table.Get(ctx, r.Rows[i].ID, r.Rows[i].Doc)
+				if err != nil {
+					return r, err
+				}
+			}
+		}
+		return r, nil
+	}
+
+	value, err := v.config.Reducer(r.Rows, false)
```

Comment:
This doesn't seem to handle querying multiple keys yet? You should panic if that's the case, and put a TODO. And the next step should implement such support.